### ISSUE TITLE
moe通信

### ISSUE THREAD

**Issue by floatlibai** [floatlibai/Blog#5]
### Reference
https://www.cnblogs.com/rossiXYZ/p/18800825#%E6%8E%A2%E7%A7%98transformer%E7%B3%BB%E5%88%97%E4%B9%8B21----moe

**Comment by floatlibai** [comment]
```mermaid
sequenceDiagram
      autonumber

      participant MR as ModelRunner
      participant CTX as ForwardContext
      participant Model as MoE Model
      participant Runner as AscendMoERunner
      participant RE as AscendRoutedExperts
      participant Router as MoE Router
      participant Comm as MoECommMethod
      participant Dispatcher as All2All Dispatcher
      participant EP as Other EP Ranks
      participant MLP as Expert MLP

      MR->>CTX: set_ascend_forward_context()
      CTX->>CTX: select_moe_comm_method()
      CTX->>CTX: moe_comm_type = ALLTOALL
      CTX->>CTX: moe_comm_method = AlltoAllCommImpl

      MR->>Model: model.forward()
      Model->>Model: router_logits = gate(hidden_states)
      Model->>Runner: self.experts(hidden_states, router_logits)

      Runner->>Runner: MoERunner.forward()
      Runner->>RE: forward_impl(hidden_states, router_logits)

      rect rgb(235, 245, 255)
          Note over RE,Comm: Prepare
          RE->>Comm: prepare(hidden_states, router_logits)
          Comm->>Comm: TP padding
          Comm->>Comm: 按 TP rank 切分 Token
          Comm-->>RE: local hidden_states, local router_logits
      end

      rect rgb(245, 240, 255)
          Note over RE,Router: Router / Top-K
          RE->>Router: _select_experts(hidden_states, router_logits)
          Router->>Router: softmax/sigmoid + Top-K
          Router->>Router: EPLB logical-to-physical mapping
          Router-->>RE: topk_ids, topk_weights
      end

      RE->>Comm: fused_experts(x, topk_ids, topk_weights)

      rect rgb(255, 245, 225)
          Note over Comm,EP: Dispatch

          Comm->>Dispatcher: token_dispatch()

          Dispatcher->>Dispatcher: 统计 tokens_per_expert
          Dispatcher->>Dispatcher: 计算 input_splits
          Dispatcher->>EP: AllGather 每个 Rank 的专家计数
          EP-->>Dispatcher: 全局专家 Token 计数
          Dispatcher->>Dispatcher: 计算 output_splits

          Dispatcher->>Dispatcher: Token 按目标 Rank/Expert 排序
          Dispatcher->>Dispatcher: 保存 reverse permutation

          Dispatcher->>EP: all_to_all_single(Token, input_splits)
          EP-->>Dispatcher: 接收属于本地专家的 Token

          Dispatcher->>Dispatcher: 按 local expert 二次排序
          Dispatcher-->>Comm: dispatched tokens + group_list + metadata
      end

      rect rgb(235, 255, 235)
          Note over Comm,MLP: Local Expert Compute

          Comm->>MLP: unified_apply_mlp(dispatched tokens)

          MLP->>MLP: GMM1: gate/up projection
          MLP->>MLP: SwiGLU/GELU activation
          MLP->>MLP: GMM2: down projection

          MLP-->>Comm: local expert outputs
      end

      rect rgb(255, 235, 235)
          Note over Comm,EP: Combine

          Comm->>Dispatcher: token_combine(expert outputs)

          Dispatcher->>Dispatcher: 撤销 local expert 排序
          Dispatcher->>EP: all_to_all_single(expert outputs)
          EP-->>Dispatcher: 返回原 Token 所在 Rank

          Dispatcher->>Dispatcher: 恢复原始 Token 顺序
          Dispatcher->>Dispatcher: Top-K 权重加权求和
          Dispatcher-->>Comm: routed_output
      end

      Comm-->>RE: FusedExpertsResult

      rect rgb(240, 240, 240)
          Note over RE,Comm: Finalize
          RE->>Comm: finalize(routed_output)
          Comm->>Comm: TP AllGather
          Comm->>Comm: 去除 padding
          Comm-->>RE: local MoE output
      end

      RE-->>Runner: routed_output

      opt 存在 Shared Expert
          Runner->>Runner: shared_expert(hidden_states)
          Runner->>Runner: routed_output + shared_output
      end

      Runner-->>Model: final_hidden_states
      Model-->>MR: hidden_states
```

**Comment by floatlibai** [comment]
```mermaid
 sequenceDiagram
    autonumber
    participant MR as ModelRunner
    participant Model as MoE Model
    participant Runner as AscendMoERunner
    participant RE as AscendRoutedExperts<br/>Rank r
    participant Router as Router
    participant D as All2All Dispatcher<br/>Rank r
    participant EP as Other EP Ranks
    participant MLP as Local Expert MLP<br/>E_local experts

    MR->>MR: set_ascend_forward_context()
    Note over MR: 根据芯片、Token 数、EP size<br/>选择 ALLTOALL
    MR->>Model: model.forward()
    Note over Model: hidden_states:<br/>[T_r, H]
    Model->>Runner: self.experts(hidden_states, router_logits)
    Runner->>RE: forward_impl()

    rect rgb(235,245,255)
        RE->>RE: moe_comm_method.prepare()
        Note over RE: 可能执行 padding / TP split / DP gather<br/>hidden_states: [T_r,H] → [N_r,H]<br/>router_logits: [T_r,E] → [N_r,E]
    END

    rect rgb(245,240,255)
        RE->>Router: _select_experts()
        Note over Router: router_logits: [N_r,E]<br/>topk_ids: [N_r,K]<br/>topk_weights: [N_r,K]
        Router-->>RE: topk_ids, topk_weights
    end

    RE->>D: fused_experts(x, topk_ids, topk_weights)
    Note over D: x: [N_r,H]<br/>需要展开为 M_r=N_r×K 个 pair

    rect rgb(255,245,225)
        D->>D: _dispatch_preprocess()
        Note over D: npu_moe_token_permute()<br/>按 topk_ids 展开并排序<br/>[N_r,H] → [M_r,H]<br/>保存 reversed_local_permutation_mapping
        D->>D: 统计 tokens_per_expert
        Note over D: tokens_per_expert: [E]<br/>input_splits: [P]<br/>表示本 Rank 发往各目标 Rank 的 pair 数
        D->>EP: Gather expert counts
        EP-->>D: 全局专家 Token 计数
        Note over D: output_splits: [P]<br/>表示各 Rank 发往本 Rank 的 pair 数
        D->>EP: all_to_all_single()<br/>输入 [M_r,H]
        EP-->>D: 接收远端 pair
        Note over D: 接收结果: [R_r,H]
    end

    rect rgb(255,235,225)
        D->>D: _dispatch_postprocess()
        Note over D: npu_moe_token_permute()<br/>按本 Rank 的 local expert 重新排序<br/>[R_r,H] → [R_r,H]<br/>顺序变为 Expert 0...Expert E_local-1
    end

    D->>MLP: unified_apply_mlp()
    Note over MLP: 输入: [R_r,H]<br/>group_list: [E_local]<br/>每个专家处理自己的 Token 子集
    MLP->>MLP: GMM1 / gate_up
    Note over MLP: [R_r,H] → [R_r,2I]
    MLP->>MLP: SwiGLU / GELU
    Note over MLP: [R_r,2I] → [R_r,I]
    MLP->>MLP: GMM2 / down_proj
    MLP-->>D: expert output: [R_r,H]

    rect rgb(255,235,235)
        D->>D: _combine_preprocess()
        Note over D: npu_moe_token_unpermute()<br/>撤销“按 local expert 排序”<br/>[R_r,H] → [R_r,H]<br/>恢复 All-to-All 前的 pair 顺序
        D->>EP: all_to_all_single()<br/>输入 [R_r,H]
        EP-->>D: 返回原 Rank 的 pair
        Note over D: 当前 Rank 收到: [M_r,H]
    end

    rect rgb(245,235,255)
        D->>D: _combine_postprocess()
        Note over D: npu_moe_token_unpermute()<br/>使用 reversed_local_permutation_mapping<br/>并传入 topk_weights [N_r,K]<br/>[M_r,H] → [N_r,H]<br/>完成 K 路专家结果加权求和
    end

    D-->>RE: routed_out [N_r,H]

    rect rgb(235,245,235)
        RE->>RE: moe_comm_method.finalize()
        Note over RE: TP gather / DP reduce-scatter / unpad<br/>[N_r,H] → [T_r,H]
    end

    RE-->>Runner: routed_output [T_r,H]

    opt 存在 Shared Expert
        Runner->>Runner: shared_output [T_r,H]
        Runner->>Runner: routed_output + shared_output
        Note over Runner: 最终输出: [T_r,H]
    end

    Runner-->>Model: final_hidden_states [T_r,H]

```

**Comment by floatlibai** [comment]
## 1. ALLGATHER

```mermaid
  flowchart TD
      A["Rank r 本地输入<br/>X_r: [T_r, H]"]
      B["AllGather<br/>聚合 DP / EP / SP 组内 Token"]
      C["每个 Rank 获得全局或扩展 Token<br/>X_global: [T, H]"]
      D["Router + TopK<br/>topk_ids: [T, K]<br/>topk_weights: [T, K]"]
      E["npu_moe_init_routing<br/>按本地专家范围 Permute"]
      F["本地专家计算<br/>Expert MLP<br/>输入: [N_local, H]"]
      G["npu_moe_token_unpermute<br/>恢复 Token 顺序并按 TopK 加权"]
      H["ReduceScatter / TP 聚合"]
      I["Rank r 输出<br/>Y_r: [T_r, H]"]

      A --> B --> C --> D --> E --> F --> G --> H --> I

      N["说明：<br/>每个 Rank 可能看到完整 Token 集合<br/>但只计算自己持有的本地专家"]
      C -.-> N
```

  核心逻辑：

  本地 Token
  → AllGather
  → 每个 Rank 获得全局 Token
  → 本地专家计算
  → Unpermute + Top-K 聚合
  → ReduceScatter

  特点是 Token 复制较多，兼容性好，但 EP 规模增大时通信和计算冗余会增加。

  ———

  ## 2. ALLTOALL

```mermaid
  flowchart TD
      A["Rank r 本地输入<br/>X_r: [T_r, H]"]
      B["Router + TopK<br/>topk_ids: [T_r, K]<br/>topk_weights: [T_r, K]"]
      C["第一次 Permute<br/>Token 展开为 pair<br/>[T_r, H] → [T_r×K, H]"]
      D["按目标 EP Rank 计算 split<br/>input_splits: [P]<br/>output_splits: [P]"]
      E["All-to-All-V Dispatch<br/>发送: [T_r×K, H]"]
      F["Rank r 接收远端 pair<br/>X_recv: [R_r, H]"]
      G["第二次 Permute<br/>按本地 Expert 排序<br/>[R_r, H] → [R_r, H]"]
      H["本地 Expert MLP<br/>Expert 0...E_local-1<br/>输出: [R_r, H]"]
      I["Unpermute<br/>撤销本地 Expert 排序"]
      J["All-to-All-V Combine<br/>返回原 Token 所在 Rank"]
      K["Rank r 收到本地 pair<br/>[T_r×K, H]"]
      L["Unpermute + TopK 加权<br/>[T_r×K, H] → [T_r, H]"]
      M["Rank r 输出<br/>Y_r: [T_r, H]"]

      A --> B --> C --> D --> E --> F --> G --> H --> I --> J --> K --> L --> M

      N["每个 Token 只发送给<br/>所选专家所在的 EP Rank"]
      E -.-> N
```

  核心逻辑：

  本地 Token
  → Permute
  → All-to-All-V
  → 按本地专家重新 Permute
  → Expert MLP
  → Unpermute
  → All-to-All-V
  → Unpermute + Top-K 加权

  这是最标准、最容易观察的 EP 路径。通信量主要与：

  T_r × K × H

  相关，通常比全量 AllGather 更节省，但需要两次 All-to-All 和多组 permutation metadata。

  ———

  ## 3. MC2

```mermaid
  flowchart TD
      A["Rank r 输入<br/>X_r: [T_r, H]"]
      B["Forward Context<br/>Padding / TP Slice / mc2_mask"]
      C["本地 Router + TopK<br/>topk_ids: [N_r, K]<br/>topk_weights: [N_r, K]"]
      D["npu_moe_distribute_dispatch[_v2]"]
      E["算子内部完成：<br/>Token 展开<br/>EP Dispatch<br/>Token Permute<br/>生成 assist_info<br/>生成 expert_token_nums"]
      F["Rank r 获得本地专家输入<br/>X_local: [R_r, H]"]
      G["本地 Expert MLP<br/>Grouped MatMul<br/>输出: [R_r, H]"]
      H["npu_moe_distribute_combine[_v2]"]
      I["算子内部完成：<br/>EP Combine<br/>Token Unpermute<br/>Top-K 加权聚合"]
      J["去 Padding / 恢复 TP 输出"]
      K["Rank r 输出<br/>Y_r: [T_r, H]"]

      A --> B --> C --> D --> E --> F --> G --> H --> I --> J --> K

      N["Python 层不直接调用<br/>dist.all_to_all_single<br/>通信和路由 metadata 由 CANN 管理"]
      D -.-> N
```

  核心逻辑：

  输入准备
  → Router + TopK
  → npu_moe_distribute_dispatch
  → 本地 Expert MLP
  → npu_moe_distribute_combine
  → 输出恢复

  MC2 的 Dispatch/Combine 是 Ascend 专用 MoE 算子，内部负责部分 Permute、通信和索引管理。通常需要固定或受限的 Token capacity，因
  此可能涉及 Padding 和 mc2_mask。

  ———

  ## 4. FUSED_MC2

```mermaid
  flowchart TD
      A["Rank r 输入<br/>X_r: [T_r, H]"]
      B["Forward Context<br/>Padding / mc2_mask"]
      C["Router 输出<br/>topk_ids: [N_r, K]<br/>topk_weights: [N_r, K]"]
      D["MegaMoe / dispatch_ffn_combine"]

      D1["内部 Token Dispatch<br/>Permute + EP 通信"]
      D2["内部 Expert GMM1<br/>Gate / Up Projection"]
      D3["内部 Activation<br/>SwiGLU / GELU"]
      D4["内部 Expert GMM2<br/>Down Projection"]
      D5["内部 Token Combine<br/>EP 通信 + Unpermute<br/>Top-K 加权"]

      E["Rank r routed output<br/>[T_r, H]"]

      A --> B --> C --> D
      D --> D1 --> D2 --> D3 --> D4 --> D5 --> E

      N["Python 层通常只看到：<br/>输入 Tensor → 输出 Tensor<br/>中间 Token 和 metadata 由融合算子管理"]
      D -.-> N
```

  核心逻辑：

  Router + TopK
  → MegaMoe / dispatch_ffn_combine
      ├─ Dispatch
      ├─ EP 通信
      ├─ Expert GMM1
      ├─ Activation
      ├─ Expert GMM2
      ├─ Combine
      └─ Top-K 聚合
  → routed output

  FUSED_MC2 将通信、Permute、Expert MLP 和 Combine 尽量融合到一个 CANN 算子中，中间形状通常不暴露给 Python 层。

  当前代码中的两个实现分支：

  A3 / MegaMoe 支持：
      self.mega_moe(...)

  旧版融合算子：
      torch.ops._C_ascend.dispatch_ffn_combine(...)

  ———

  ## 四种模式的核心区别

```mermaid
  flowchart LR
      A["ALLGATHER<br/>先聚合 Token"]
      B["ALLTOALL<br/>按目标专家发送 Token"]
      C["MC2<br/>用 CANN 算子完成 Dispatch/Combine"]
      D["FUSED_MC2<br/>通信和专家计算进一步融合"]

      A --> B --> C --> D
```

   模式                 ALLGATHER
   Token 通信           AllGather + ReduceScatter
   Permute              部分由 NPU routing 算子完成
   Expert MLP           独立执行
   Python 可见中间过程  较高
   主要取舍             兼容性好，但 Token 复制多
  ───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   模式                 ALLTOALL
   Token 通信           两次 All-to-All-V
   Permute              显式 token_permute/unpermute
   Expert MLP           独立 GMM
   Python 可见中间过程  最高
   主要取舍             通用 EP，通信冗余较少
  ───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   模式                 MC2
   Token 通信           CANN Dispatch/Combine
   Permute              CANN 内部完成
   Expert MLP           通常独立 GMM
   Python 可见中间过程  中等
   主要取舍             Ascend 专用优化
  ───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   模式                 FUSED_MC2
   Token 通信           融合算子内部完成
   Permute              基本隐藏
   Expert MLP           与通信融合
   Python 可见中间过程  最低
   主要取舍             性能高，但约束和依赖最多

  对应代码实现：

  ALLGATHER:
    moe_comm_method.py:176
    token_dispatcher.py:358

  ALLTOALL:
    moe_comm_method.py:226
    token_dispatcher.py:464

  MC2:
    moe_comm_method.py:206
    token_dispatcher.py:113

  FUSED_MC2:
    moe_comm_method.py:250
    moe_comm_method.py:412

**Comment by floatlibai** [comment]
fused_mc2
当前vllm-ascend支持配置enable_fused_mc2开启mc2的大融合算子，讲解清楚vllm-ascend完整的调用链，以及dispatchh_ffn_combine和megamoe的核心逻辑。
另外，enable_mc2_hierarchy_comm和enable_prefill_mc2也与mc融合相关，一起讲解

参考：https://github.com/Ronald1995/infra/blob/main/vllm-ascend/enable_fused_mc2.md
https://github.com/Ronald1995/infra/blob/main/vllm-ascend/enable_prefill_mc2.md
https://github.com/Ronald1995/infra/blob/main/vllm-ascend/enable_mc2_hierarchy_comm.md
